inject/src/main/java/io/micronaut/context/LazyProvider.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.context;
+
+import javax.inject.Provider;
+
+/**
+ * A lazy provider.
+ *
+ * @param <T> The type
+ * @author Denis Stepanov
+ * @since 2.0.2
+ */
+abstract class LazyProvider<T> implements Provider<T> {
```

Comment:
Maybe instead of introducing a new class `SupplierUtil.memoized` could be used instead?

Reply:
It can be, I didn’t know about it, Provider might be useful for direct injection of it.
Do you want to use Supplier instead?

Reply:
@dstepanov If you want to create a `ProviderUtil.memoized` that would be fine. Either way the synchronization logic should be externalized for reusability. The supplier one works well so use that as an example

Reply:
I think an issue there may be `Provider` is part of `java.inject` and not available in the `core` module

Reply:
I have created an util class in the same package.

Reply:
@jameskleeh What do you think about the latest version?

Reply:
@dstepanov The util doesn't follow the pattern of SupplierUtil. The crux of the implementation is the delegate gets overridden so it doesn't use synchronization after the first invocation. You can message me on gitter if you'd like more details

Reply:
@jameskleeh Changed